### PR TITLE
Weight compression via Lora Correction Algorithm

### PR DESCRIPTION
**Pull request by ljaljushkin**
### Changes

Lora Correction algorithm for int4/nf4 weight compression. 

### Reason for changes

Method for improving accuracy by migrating quantization noise to “learnable” lora adapters.

### Related tickets

135863

### Tests

- [x] docstrings, proper names
- [x] results for phi3 and stablelm2-1.6b on lambada, wikitext
- [x] job/NNCF/job/manual/job/post_training_weight_compression/144/
image
image
image

image

### CODE REVIEW

File: nncf/quantization/algorithms/weight_compression/lora.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                self._activations,
+                self._debug_interface,
+            )
+        finally:
+            if self._debug_interface is not None:
+                self._debug_interface.dump_data()
+        return adapters
+
+    @staticmethod
+    def calculate_adapters_fn(
```
Comment (ljaljushkin) [inline nncf/quantization/algorithms/weight_compression/lora.py:98]: The main functionality of the algorithm is here. @andreyanufr 
Designed to be called from other places, e.g. GPTQ @alexsu52

File: nncf/quantization/algorithms/weight_compression/algorithm.py [hunk 2]
```diff
@@ -403,6 +407,11 @@ def apply(
                 backend_entity=self._backend_entity,
             )
 
+        lora_algorithm = None
+        if self._lora_correction:
+            lora_params = self._advanced_parameters.lora_params
+            lora_algorithm = LoraCorrectionAlgorithm(activations, lora_params)
```
Comment (ljaljushkin) [inline nncf/quantization/algorithms/weight_compression/algorithm.py:413]: @alexsu52 
Introduced `LoraAlgorithm` wrapper just for simplicity. The algorithm can be applied to the layer without it.

File: nncf/quantization/algorithms/weight_compression/activation_stats.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# limitations under the License.
+
+from typing import List, Tuple, TypeVar
+
+from nncf.tensor import functions as fns
+
+TTensor = TypeVar("TTensor")
+
+
+def process_stats(stats: List[TTensor], subset_size: int) -> Tuple[TTensor, TTensor]:
```
Comment (ljaljushkin) [inline nncf/quantization/algorithms/weight_compression/activation_stats.py:19]: @andreyanufr, here's the common function for AWQ, Scale Estimation and Lora Correction

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py [hunk 4]
```diff
@@ -171,55 +261,24 @@ def transform_model(
...
-                    mul, const_dtype, name=f"{const_node_name}/fq_weights_{wc_params.weight_port_id}/convert"
-                )
-
             mul_output = mul.output(0)
             for target_input in const_node.output(0).get_target_inputs():
                 target_input.replace_source_output(mul_output)
 
+            if lora_algorithm is not None and lora_algorithm.is_applicable(wc_params):
+                V, US = lora_algorithm.calculate_adapters(weight, compressed_weight, wc_params)
+                self.insert_adapters(wc_params, V, US, int8_lora=lora_algorithm.is_int8_adapters)
```
Comment (ljaljushkin) [inline nncf/quantization/algorithms/weight_compression/openvino_backend.py:280]: Lora Correction Algorithm is applied layer-by-layer after creating a compressed weight.

File: nncf/quantization/algorithms/weight_compression/algorithm.py [hunk 5]
```diff
@@ -403,16 +407,24 @@ def apply(
...
 
         # Compress model using weight compression parameters
         transformed_model = self._backend_entity.transform_model(
             model,
             graph,
-            track(all_weight_params, description="Applying Weight Compression"),
+            track(all_weight_params, description=description),
             scales,
             zero_points,
+            lora_correction_algo,
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/algorithm.py:427]: From a design point of view, you pass backend agnostic algorithm to the backend specific function. I think, we can avoid such behavior if the transformation model loop will be implemented in the backend agnostc code and the algo backend will have only the atomic compression transformation (per weight). Thus you can call lora correction algorithm into the backend agnostic code, for example in the compression algorithm.

Note: I recommend to refactor this part, but this is not a stopper for merge and can be done in a follow-up PR.

File: nncf/quantization/algorithms/weight_compression/backend.py [hunk 6]
```diff
@@ -127,6 +127,36 @@ def transform_model(
...
+                   orig.MM--------------------------/
+
+        :param wc_params: Parameters for weight compression.
+        :type wc_params: WeightCompressionParameters
+        :param lora_A: weights for the first LoRA matrix.
+        :type lora_A: Tensor
+        :param lora_B: weights for the second LoRA matrix.
+        :type lora_B: Tensor
+        :param int8_lora: indicates whether the LoRA matrices should be compressed to 8-bit.
+        :type int8_lora: bool
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/backend.py:157]:
```suggestion
        :param wc_params: Parameters for weight compression.
        :param lora_A: weights for the first LoRA matrix.
        :param lora_B: weights for the second LoRA matrix.
        :param int8_lora: indicates whether the LoRA matrices should be compressed to 8-bit.
```

File: nncf/quantization/algorithms/weight_compression/backend.py [hunk 7]
```diff
@@ -127,6 +127,36 @@ def transform_model(
         :return: The transformed model.
         """
 
+    @abstractmethod
+    def insert_adapters(self, wc_params: WeightCompressionParameters, lora_A: Tensor, lora_B: Tensor, int8_lora: bool):
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/backend.py:131]: Please, add return type hint.

File: nncf/quantization/algorithms/weight_compression/lora_correction.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+
+class LoraCorrectionAlgorithm:
+    """
+    Contains implementation of LoRA Correction algorithm.
+
+    The method reduces quantization noise after weight compression using low rank adapters.
+    """
+
+    def __init__(self, activations: Dict[str, List[Tensor]], lora_correction_params: AdvancedLoraCorrectionParameters):
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/lora_correction.py:81]: What do you think about explicit definition of algorithm parameters?

The current implementation: 
```
from nncf.quantization.advanced_parameters import AdvancedLoraCorrectionParameters

algo = LoraCorrectionAlgorithm(
    activations=activations,
    lora_correction_params=AdvancedLoraCorrectionParameters(
        rank=8
        n_iters=3
        w_regularization=False
        subset_size=32
        is_int8_adapters=True
    )
)
```
or 
```
algo = LoraCorrectionAlgorithm(
    activations=activations,
    rank=8
    n_iters=3
    w_regularization=False
    subset_size=32
    is_int8_adapters=True
)
```

File: nncf/quantization/algorithms/weight_compression/lora_correction.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+                self._activations,
+                self._debug_interface,
+            )
+        finally:
+            if self._debug_interface is not None:
+                self._debug_interface.dump_data()
+        return adapters
+
+    @staticmethod
+    def calculate_adapters_fn(
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/lora_correction.py:123]:
```suggestion
    def calculate_low_rank_matrices(
```

File: nncf/quantization/algorithms/weight_compression/lora_correction.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+        weight: Tensor,
+        compressed_weight: Tensor,
+        wc_params: WeightCompressionParameters,
+        lora_correction_params: AdvancedLoraCorrectionParameters,
+        activations: Dict[str, List[Tensor]],
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/lora_correction.py:128]: I would suggest to feed only those activations that the function needs for calculations.

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py [hunk 11]
```diff
@@ -131,38 +133,128 @@ def set_weight(
...
 
+    def insert_adapters(self, wc_params: WeightCompressionParameters, lora_A: Tensor, lora_B: Tensor, int8_lora: bool):
+        input_node = self.name_to_node_mapping[wc_params.node_with_weight.node_name].input_value(0)
+        activation_dtype = input_node.get_element_type()
+        should_add_convert_node = activation_dtype != ov.Type.f16
+        mm_node = self.name_to_node_mapping[wc_params.node_with_weight.node_name]
+
+        if int8_lora:
+            a_wc_params = deepcopy(wc_params)
+            a_wc_params.compression_config = WeightCompressionConfig()
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/openvino_backend.py:144]: I believe this is a workaround for the bad function signature of `_create_compression_subgraph`.

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py [hunk 12]
```diff
@@ -131,38 +133,128 @@ def set_weight(
...
+                wc_params=a_wc_params,
+                const_node_name=a_wc_params.node_with_weight.node_name + "_lora_B",
+                const_dtype=activation_dtype,
+                should_add_convert_node=should_add_convert_node,
+            )
+        else:
+            A_W = opset.constant(lora_A.data)
+            B_W = opset.constant(lora_B.data)
+
+        A_MM = opset.matmul(input_node, A_W, transpose_a=False, transpose_b=True)
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/openvino_backend.py:164]: The `_create_compression_subgraph` uses `weight_port_id` from `a_wc_params`, but weight_port_id for lora adapters always is 1.

File: nncf/quantization/algorithms/weight_compression/weight_lowering.py [hunk 13]
```diff
@@ -171,6 +153,29 @@ def calculate_normalized_weight(weight: Tensor, scale: Tensor) -> Tensor:
...
+    """
+    Does NF4 quantization and dequantization (fake quantization) for the weights normalized by the given scale.
+
+    :param norm_weight: weight divided by the given scale.
+    :param scale: Scale tensor used for normalization.
+    :param reduction_axis: axis along which weights were reshaped for group quantization and will be reshaped back to
+        original shapes. If equals to -1: weights are not reshaped, assumed not a group quantization. Default to -1.
+    :return: NF4 fake-quantized weight tensor.
+    """
+    center_nf4_quantiles = fns.from_numpy(CENTER_OF_NF4_QUANTILES, backend=norm_weight.backend)
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/weight_lowering.py:166]: Please remove copy-past from this function.

File: nncf/quantization/algorithms/weight_compression/lora.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+        Vr = Sr @ Vr
+        US = Ur
+
+        dY = w_residual @ X
+
+        for i in range(n_iters):
+            VX = Vr @ X
+            if not w_regularization:
+                sol = fns.linalg.lstsq(fns.transpose(VX), fns.transpose(dY))
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/quantization/algorithms/weight_compression/lora.py]: thanks, corrected

File: nncf/quantization/algorithms/weight_compression/lora_correction.py [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+        if compression_config.group_size > 0:
+            # Multiply residual of weights by maximum channel magnitude of activations normalized per quantization
+            # group. As a consequence, weights corresponding to a "noisy" activations has a higher error to correct.
+            # Empirically, it leads to a better accuracy.
+            gs = compression_config.group_size
+            n_gs = s.shape[0] // gs
+            for i in range(n_gs):
+                offset = i * gs
+                denum = fns.sum(s[offset : offset + gs])
+                s[offset : offset + gs] = s[offset : offset + gs] / denum
```
Comment (andreyanufr) [inline nncf/quantization/algorithms/weight_compression/lora_correction.py:202]: I think we can skip first normalization

**Review by andreyanufr — APPROVED** [review]

File: nncf/quantization/quantize_model.py [hunk 16]
```diff
@@ -496,12 +504,18 @@ def compress_weights(
...
+            "sensitivity_metric": sensitivity_metric,
+            "dataset": dataset,
+            "awq": awq,
+            "scale_estimation": scale_estimation,
+            "gptq": gptq,
+            "lora_correction": lora_correction,
+        }
+        unsupported_for_int8 = [name for name, value in options.items() if value is not None]
+        if unsupported_for_int8:
+            raise AttributeError(f"INT8 modes do not support {unsupported_for_int8} option(s). Set them to None.")
```
Comment (alexsu52) [inline nncf/quantization/quantize_model.py:518]:
```suggestion
            raise AttributeError(f"INT8 modes do not support {', '.join(unsupported_for_int8)} option(s). Set them to None.")
```

**Review by alexsu52 — COMMENTED** [review]
LGTM, please consider taking my comments into consideration

File: nncf/quantization/advanced_parameters.py [hunk 17]
```diff
@@ -314,6 +314,33 @@ class AdvancedGPTQParameters:
...
+    :param is_int8_adapters: Whether to 8-bit quantize lora adapters, otherwise they kept in the original weights
+        precision. Defaults to True.
+    :type is_int8_adapters: bool
+    """
+
+    rank: int = 8
+    num_iters: int = 3
+    add_regularization: bool = True
+    subset_size: int = 128
+    is_int8_adapters: bool = True
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/quantization/advanced_parameters.py]: Thanks, corrected

File: nncf/quantization/quantize_model.py (hunk 16 above)
Comment (ljaljushkin) [reply nncf/quantization/quantize_model.py]: Applied the change

**Comment by ljaljushkin** [comment]
Latest build of conformance test with lora test case:
job/NNCF/job/manual/job/post_training_weight_compression/150

**Review by alexsu52 — APPROVED** [review]

**Pull request merged**